### PR TITLE
samples: nrf_desktop: Bond operations with single button

### PR DESCRIPTION
**Pull request by pdunaj**
Jira:DESK-418

Signed-off-by: Pawel Dunaj <pawel.dunaj@nordicsemi.no>

### CODE REVIEW

**Comment by pdunaj** [comment]
I think I found a way to work around the fact that we don't have whitelisting yet... I will send one update in few minutes if it works.

**Comment by Qbicz** [comment]
Can you push it as a separate commit to squash later? I'm already in the middle of the review ;)

File: samples/nrf_desktop/src/modules/ble_state.c [hunk 1]
```diff
@@ -19,45 +19,86 @@
...
 
 	int err = bt_conn_get_info(conn, &info);
 	if (err) {
-		LOG_ERR("Cannot get conn info");
-		return;
+		LOG_WRN("Cannot get conn info");
+		goto disconnect;
 	}
 
+	struct bond_find_data bond_find_data = {
```
Comment (pdunaj) [inline samples/nrf_desktop/src/modules/ble_state.c]: do bond check only for peripheral device

File: samples/nrf_desktop/src/modules/ble_bond.c [hunk 2]
```diff
@@ -10,78 +10,367 @@
...
-	return 0;
+	cancel_operation();
 }
 
-static int cmd_bonds_remove(const struct shell *shell, size_t argc, char **argv)
+static void single_click_handler(struct k_work *work)
 {
-	int err = bonds_remove();
+	handle_click(CLICK_SHORT);
+	single_click.work_q = NULL;
```
Comment (Qbicz) [inline samples/nrf_desktop/src/modules/ble_bond.c]: Why do you need to set work_q to NULL?

File: samples/nrf_desktop/src/modules/ble_bond.c [hunk 3]
```diff
@@ -10,78 +10,367 @@
...
+#include "ble_event.h"
 
 #include <logging/log.h>
 LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BLE_BOND_LOG_LEVEL);
 
-static bool bonds_initialized;
-static bool remove;
 
-static int bonds_remove(void)
+#define SHORT_CLICK_MAX 500 /* ms */
```
Comment (Qbicz) [inline samples/nrf_desktop/src/modules/ble_bond.c]: `K_MSEC(500)` and you can remove comment

**Comment by pdunaj** [comment]
The change works but I see some issues when connecting, turning BLE on host off, attempting to connect from some other host and then BLE back on on the original host. There are some errors in the stack. I will track it but I would prefer to fix it as separate PR unless change is trivial. I don't want to complicate this one more then necessary.

File: samples/nrf_desktop/src/modules/ble_bond.c [hunk 4]
```diff
@@ -10,78 +10,367 @@
...
-	shell_print(shell, "\t%s", addr);
+static void handle_click(enum click click)
+{
+	if (click != CLICK_NONE) {
+		for (size_t i = 0; i < ARRAY_SIZE(state_switch); i++) {
+			if ((state_switch[i].state == state) &&
+			    (state_switch[i].click == click)) {
+				if (NULL != state_switch[i].func) {
+					state_switch[i].func();
+				}
```
Comment (Qbicz) [inline samples/nrf_desktop/src/modules/ble_bond.c]: ASSERT or `else {LOG_ERR("Callback missing")}`

File: samples/nrf_desktop/src/modules/ble_bond.c [hunk 5]
```diff
@@ -10,78 +10,367 @@
...
-SHELL_STATIC_SUBCMD_SET_CREATE(sub_bonds,
-	SHELL_CMD_ARG(show, NULL, "Show bonds", cmd_bonds_show, 0, 0),
-	SHELL_CMD_ARG(remove, NULL, "Remove all bonded devices",
-		      cmd_bonds_remove, 0, 0),
-	SHELL_SUBCMD_SET_END
-);
+static void load_identities(void)
+{
+	bt_addr_le_t addrs[CONFIG_BT_ID_MAX];
+	size_t count;
```
Comment (Qbicz) [inline samples/nrf_desktop/src/modules/ble_bond.c]: `count` Should be initialized to the `addrs` array size before calling `bt_id_get`

File: samples/nrf_desktop/src/modules/ble_bond.c (hunk 2 above)
Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_bond.c]: To detect that work was executed when I call cancel.

File: samples/nrf_desktop/src/modules/ble_bond.c (hunk 4 above)
Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_bond.c]: My assumption was that it is perfectly fine not to have this callback. It's when you only want to change state without any effect.

File: samples/nrf_desktop/src/modules/ble_bond.c (hunk 5 above)
Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_bond.c]: I guess it does not matter really as documentation states
`the number of elements in the @a addrs array should be CONFIG_BT_ID_MAX`
but later it states `count Should be initialized to the array size` so I will set it

File: samples/nrf_desktop/src/modules/ble_bond.c (hunk 4 above)
Comment (Qbicz) [reply samples/nrf_desktop/src/modules/ble_bond.c]: Agree, it's a state implementation using a state matrix, could be without callback.

File: samples/nrf_desktop/src/modules/ble_bond.c (hunk 5 above)
Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_bond.c]: Fixed

File: samples/nrf_desktop/src/modules/ble_bond.c (hunk 3 above)
Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_bond.c]: Fixed

File: samples/nrf_desktop/src/modules/ble_bond.c (hunk 2 above)
Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_bond.c]: Added a comment

**Comment by pdunaj** [comment]
Rebased. Fixed macro name. Fixed bond check only on peripheral.
@Qbicz you can have a look in ble_state if you like :)

File: samples/nrf_desktop/src/modules/ble_adv.c [hunk 6]
```diff
@@ -95,6 +95,7 @@ static enum state state;
 
 static struct k_delayed_work adv_update;
 static struct k_delayed_work sp_grace_period_to;
+static u8_t cur_identity = BT_ID_DEFAULT; /* We expect zero */
```
Comment (MarekPieta) [inline samples/nrf_desktop/src/modules/ble_adv.c]: Missing dot ?

Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_adv.c]: on purpose

Comment (pdunaj) [reply samples/nrf_desktop/src/modules/ble_adv.c]: treated as short comment instead of sentence. should I add it?

**Comment by pdunaj** [comment]
@pizi-nordic , could you merge?

**Pull request merged**